Simplified testing system

smharper opened the pull request:
This PR introduces a simpler, object-oriented testing system.  All of the "results.py" files are gone.  Most of the "test_*.py" files now simply import the new `TestHarness` class, and call it's `execute_test()` method.  The more specialized tests create their own classes that inherit from `TestHarness` to check for different files, create different outputs, run OpenMC multiple times, etc.

I changed a number of the "results_true.dat" files.  Most of them simply replace "tallies:" with "tally 1:".  A few of them have other cosmetic changes like removing a newline or changing the order of results.  The only test I changed significantly was "test_filter_distribcell".  That test now creates three separate "results_test.dat" files and checks them all.  The case-3 results are a hash rather than a raw output (it would be > 100,000 lines otherwise).

wbinventor commented on tests/test_filter_cell/results_true.dat:
```diff
@@ -1,6 +1,6 @@
 k-combined:
 1.093844E+00 1.626801E-02
-tallies:
+tally 1:
```
Any reason why the tally id was inserted here?

wbinventor commented on tests/test_filter_cellborn/results_true.dat:
```diff
@@ -1,6 +1,6 @@
 k-combined:
 1.093844E+00 1.626801E-02
-tallies:
+tally 1:
```
oh, it looks like it was added everywhere so I assume the general ``TestHarness`` class must insert it now. Disregard my question about it.

wbinventor commented on tests/test_filter_distribcell/case-1/results_true.dat:
```diff
@@ -0,0 +1,14 @@
+k-combined:
```
So why did the results for cases 1 and 2 change?

smharper replied:
It looks like all the same values just placed in a different order.  If you want, I'll try to work out what the exact difference is.

wbinventor commented on tests/test_filter_distribcell/test_filter_distribcell.py:
```diff
@@ -1,109 +1,119 @@
+                    outstr += '\n'.join(results) + '\n'
+                    tally_num += 1
+
+            if i == 2:
```
Make this conditional more transparent, something like ``dirs[i] == 'case-3'``

smharper replied:
I've rewritten this test to something that I think is more clear.

wbinventor commented on tests/test_filter_distribcell/test_filter_distribcell.py:
```diff
@@ -1,109 +1,119 @@
+
+
+    def _compare_results(self):
+        dirs = ('case-1', '../case-2', '../case-3')
```
perhaps make ``dirs`` a class attribute or global variable for this test to avoid code duplication?

wbinventor commented on tests/test_plot_mask/test_plot_mask.py:
```diff
@@ -1,47 +1,47 @@
+from testing_harness import *
+
+
+class PlotTestHarness(TestHarness):
```
Perhaps a few of these more custom classes, like ``PlotTestHarness``, could be made as their own class to avoid code duplication? Would also make them importable for new tests someday.

smharper replied:
I've moved the `PlotTestHarness`, `CMFDTestHarness`, and `ParticleRestartTestHarness` classes into testing_harness.py.

wbinventor commented on tests/test_resonance_scattering/test_resonance_scattering.py:
```diff
@@ -1,60 +1,10 @@
-import filecmp
-import glob
-from optparse import OptionParser
+sys.path.insert(0, '..')
```
Why does ``sys.path.insert(0, '..')`` need to be here and elsewhere? Looks like it is already in "testing_harness.py"

smharper replied:
The insertion adds `testing_harness` to the path.  `import testing_harness` will fail without that line.

wbinventor commented on tests/test_source_file/test_source_file.py:
```diff
@@ -1,119 +1,97 @@
 </settings>
 """
 
 settings2 = """<?xml version="1.0"?>
```
Why is this XML code here?

smharper replied:
We need to run OpenMC with two different inputs.  The first creates a source_point file.  The second loads that source_point for the initial source.

wbinventor commented on tests/test_trigger_batch_interval/results_true.dat:
```diff
@@ -1,27 +1,28 @@
 k-combined:
 9.851940E-01 4.283951E-03
-tallies:
+tally 1:
```
Why did the results change?

smharper replied:
All the same numbers are there, they've just been interleaved in a different order.

wbinventor commented on tests/test_trigger_no_batch_interval/results_true.dat:
```diff
@@ -1,27 +1,28 @@
 k-combined:
 9.851940E-01 4.283951E-03
-tallies:
+tally 1:
```
ditto

smharper replied:
Same numbers, different order

wbinventor commented on tests/test_trigger_no_status/results_true.dat:
```diff
@@ -1,27 +1,28 @@
 k-combined:
 9.809303E-01 7.264435E-03
-tallies:
+tally 1:
```
ditto results changed

smharper replied:
Same numbers, different order.  Do you mind that they are in a different order?  I certainly haven't changed any OpenMC physics, and I think the test is still effective.  If you disagree, I could make a class specialization for the trigger tests.

paulromano replied:
Personally I'm ok with different order if it means avoiding a class specialization just to preserve the current ordering.

wbinventor commented on tests/testing_harness.py:
```diff
@@ -0,0 +1,116 @@
+
+
+    def _run_openmc(self):
+        if self._opts.mpi_exec != '':
```
Why not use the ``openmoc.Executor`` class here and elsewhere where you run OpenMC? It should further simplify the code and allow for greater code reuse (though you will need to modify ``Executor`` to be able to set the executable path).

wbinventor commented:
@smharper this PR is great, I like it very much. Really great way to abstract the test suite!

paulromano commented on tests/testing_harness.py:
```diff
@@ -0,0 +1,116 @@
+        self._opts = None
+        self._args = None
+
+
```
Single blank line between method definitions per PEP8.

smharper replied:
Fixed

paulromano commented on tests/testing_harness.py:
```diff
@@ -0,0 +1,116 @@
+from openmc.statepoint import StatePoint
+
+
+class TestHarness():
```
Make this a new-style class by subclassing `object`

smharper replied:
Fixed

paulromano commented:
Could you also check results for test_score_current using a hash too? I see that the file size is about 2.7 MB.

paulromano commented:
Awesome pull request. Remove 7000+ lines of redundant code? :+1:

smharper commented:
I've moved test_score_current to a hashed test.  I'd like to also make the plot tests use hash comparisons before this PR is merged.  I haven't switched over to openmc.executor yet.  I'll try to make those two changes (openmc.executor and plot hashes) in the next commit.

smharper commented:
I just realized that I messed up the distribcell case 3 (doesn't include tallies).  I'll fix that after I've had some sleep

smharper commented:
I fixed the distribcell case 3 thing.  The plot tests now compare against hashed results.  All tests now use the Python API's Executor class.  In order to use the Executor, I modified it slightly so that it returns the return code.

Using Executor had a few side-effects.  You can no longer need to pass in the name of the OpenMC executable (the --exe flag) or the MPI executable (--mpi-exec is now just a True/False flag).  CMakeLists has been modified accordingly.

Another small side-effect is that test_statepoint_restart used to run two tests, one with `-r` and one with `--restart`, but that distinction is handled by Executor now.

paulromano commented:
Looks like you accidentally added a binary .swp file. Also, you need to merge in develop and resolve any conflicts.

paulromano commented on tests/testing_harness.py:
```diff
@@ -0,0 +1,245 @@
+
+import numpy as np
+
+sys.path.insert(0, '../..')
```
I think this should be `sys.path.insert(0, '..')`

smharper replied:
When a test_*.py file imports `TestHarness`, it's cwd is the test_* directory.  So we need both '..'s.

paulromano commented on tests/testing_harness.py:
```diff
@@ -0,0 +1,245 @@
+
+    def _run_openmc(self):
+        if self._opts.mpi_exec:
+            mpi_procs = int(self._opts.mpi_np)
```
We should avoid the cast to int by just reading in the option as an int in the first place, i.e., in _parse_args you'd have:
```python
parser.add_option('--mpi_np', dest='mpi_np', type=int, default=3)
```

smharper replied:
fixed

paulromano commented:
Now that two tests have results files that are hashes of the actual results, does update_results.py need to be changed? In other words, how does one update results when they change now?

smharper commented:
I had forgotten about the update_results.py.  I modified `TestHarness` to include a method for updating results.  Using that, update_results.py seems to be working.

paulromano commented:
It looks like using Executor causes everything to fail because there's no way to specify the openmc executable.

smharper commented:
So how do we add OpenMC to the path?  Should we modify the ctest script in "run_tests.py" with something like this?
```cmake
 SET (CTEST_ENVIRONMENT
     PATH={build_dir}/bin:$PATH)
```

paulromano commented:
To me, it's really a deficiency of Executor. Someone using that class should be able to specify the path to OpenMC. Do you agree @wbinventor?

wbinventor commented:
Yes, agreed, it should be capable of handling that. I'm not around my computer at the moment to propose a solution, but can later if one hasn't already been worked out.

smharper commented:
I've tried rerunning Travis a bunch of times, maybe half a dozen, and I can't get travis_install.sh to return good for the Python 3 build.  I think we may have a real error here...

paulromano commented:
Just sent you a pull request to fix the problem.

smharper commented:
The Python 3 build failed again on this PR, but it passed over on my end, so I think we can pretend this one passed ;) https://travis-ci.org/smharper/openmc/builds/69573112

smharper commented:
Oh nevermind.  Just realized that we do different tests for PRs and commits.

paulromano commented:
@wbinventor Since you were reviewing too, did you have anything else you wanted to see addressed before merging? I'm ready to merge when you are.

wbinventor commented:
It looks good to me!
On Jul 4, 2015 6:41 PM, "Paul Romano" <notifications@github.com> wrote:

> @wbinventor <https://github.com/wbinventor> Since you were reviewing too,
> did you have anything else you wanted to see addressed before merging? I'm
> ready to merge when you are.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/402#issuecomment-118569971>.
>

smharper commented:
Thanks for the thorough review guys!  After checking 200 diffs several times over my eyes glazed a bit, and I needed some sharp independent review.
